### PR TITLE
Lowpass filter chip

### PR DESCRIPTION
**Pull request by cm0x4D**
- [ ] ~~Use unity gain for Kaiser window~~
- [ ] ~~Find parameters to reduce oscillating on rectangular signals~~
- [x] Icon
- [x] Description
- [x] Save data to NBT
- [x] Logarithmic slider for frequency
- [x] Damn I always forget the recipe 🙄 

All other filter types might not make that much of a sense with simulation frequencies that low.

---
This change is [Reviewable](https://reviewable.io/reviews/electrical-age/electricalage/638)

### CODE REVIEW

**Review by AMIDIBOSS — APPROVED** [review]

**Comment by Baughn** [comment]
On the face of it this looks good, but it's complicated. I'm going to try it on Erisia for a day or two, see if I can find a good use case and test it properly.

**Comment by Baughn** [comment]
They seem to work: https://imgur.com/FVHNjfB.png

File: src/main/java/mods/eln/gui/GuiHelper.java [hunk 1]
```diff
@@ -75,6 +75,13 @@ public GuiVerticalProgressBar newGuiVerticalProgressBar(int x, int y, int width,
         return o;
     }
 
+    public GuiVerticalCustomValuesBar newGuiVerticalCustomValuesBar(int x, int y, int width, int height, Float[] values) {
```
Comment (Baughn) [inline src/main/java/mods/eln/gui/GuiHelper.java:78]: I'll be so happy when we can replace the GUI code with LibLib.

File: src/main/java/mods/eln/sixnode/AnalogChips.kt [hunk 2]
```diff
@@ -685,3 +683,121 @@ class SampleAndHold : AnalogFunction() {
...
+    private var output = 0.0
+
+    override fun process(inputs: Array<Double?>, deltaTime: Double): Double {
+        output = Utils.limit(output + ((inputs[0] ?: 0.0) - output) * feedback * deltaTime, 0.0, 50.0)
+        return output
+    }
+
+    override fun readFromNBT(nbt: NBTTagCompound?, str: String?) {
+        nbt?.apply {
+            feedback = getDouble("feedback")
```
Comment (Baughn) [inline src/main/java/mods/eln/sixnode/AnalogChips.kt:702]: It's very unlikely that you'll need the precision, and floats are smaller. This matters when persisting, or especially for network transfer.

File: src/main/java/mods/eln/sixnode/AnalogChips.kt [hunk 3]
```diff
@@ -685,3 +683,121 @@ class SampleAndHold : AnalogFunction() {
...
+            setDouble("output", output)
+        }
+    }
+}
+
+class FilterElement(node: SixNode, side: Direction, sixNodeDescriptor: SixNodeDescriptor) :
+    AnalogChipElement(node, side, sixNodeDescriptor) {
+
+    companion object {
+        val CutOffFrequencyChangedEvents = 1
```
Comment (Baughn) [inline src/main/java/mods/eln/sixnode/AnalogChips.kt]: If this is meant to be an enum, then use an enum. See https://kotlinlang.org/docs/reference/enum-classes.html

File: src/main/java/mods/eln/gui/GuiHelper.java (hunk 1 above)
Comment (cm0x4D) [reply src/main/java/mods/eln/gui/GuiHelper.java:78]: Do not know that lib, where can I find infos about?

File: src/main/java/mods/eln/sixnode/AnalogChips.kt (hunk 2 above)
Comment (cm0x4D) [reply src/main/java/mods/eln/sixnode/AnalogChips.kt:702]: Everything on the simulation side is double. I convert it to/from float when serializing to/from client.

File: src/main/java/mods/eln/sixnode/AnalogChips.kt (hunk 3 above)
Comment (cm0x4D) [reply src/main/java/mods/eln/sixnode/AnalogChips.kt]: Imho enums are completely broken in Kotlin the same way they are in Java.

Comment (cm0x4D) [reply src/main/java/mods/eln/sixnode/AnalogChips.kt]: Kotlin 1.1 adds the method `enumValueOf<T>()` which would make enums less a pain to use.

File: src/main/java/mods/eln/gui/GuiHelper.java (hunk 1 above)
Comment (Baughn) [reply src/main/java/mods/eln/gui/GuiHelper.java:78]: There's a limited amount of documentation, unfortunately. I've been browsing the code a bit, and talking to other devs. But: https://github.com/TeamWizardry/LibrarianLib/wiki/GUI-API-Home

1.10 only thus far. There is an 1.7 version, but it's different enough that it isn't worth it.

File: src/main/java/mods/eln/sixnode/AnalogChips.kt (hunk 3 above)
Comment (Baughn) [reply src/main/java/mods/eln/sixnode/AnalogChips.kt]: Mm. Well, the 1.1 RC seems stable enough that you can switch to it if you like.

Comment (cm0x4D) [reply src/main/java/mods/eln/sixnode/AnalogChips.kt]: Oh, I just saw 1.1 is actually out ^^

**Review by AMIDIBOSS — APPROVED** [review]

**Comment by cm0x4D** [comment]
@Baughn is there still something open with this PR or can I merge?

**Review by Baughn — APPROVED** [review]
Oops, guess I didn't approve.

You can merge. But note that the "Character.valueOf" calls are redundant; you can use plain old characters.

**Pull request merged**